e2e-tests/test/anvil-apis.test.ts

```diff
@@ -0,0 +1,190 @@
+
+    // Assert
+    const balance = await userWallet.getBalance();
+    expect(balance.eq(newBalance)).to.true;
+  });
+});
+
+describe("anvil_setNonce", function () {
+  it("Should update the nonce of an account", async function () {
+    // Arrange
+    const userWallet = Wallet.createRandom().connect(provider);
+    const newNonce = 42;
+
+    // Act
+    await provider.send("anvil_setNonce", [userWallet.address, ethers.utils.hexlify(newNonce)]);
+
+    // Assert
+    const nonce = await userWallet.getNonce();
+    expect(nonce).to.equal(newNonce);
+  });
+});
+
+describe("anvil_mine", function () {
+  it("Should mine multiple blocks with a given interval", async function () {
+    // Arrange
+    const numberOfBlocks = 100;
+    const intervalInSeconds = 60;
+    const startingBlock = await provider.getBlock("latest");
+    const startingTimestamp: number = await provider.send("config_getCurrentTimestamp", []);
+
+    // Act
+    await provider.send("anvil_mine", [ethers.utils.hexlify(numberOfBlocks), ethers.utils.hexlify(intervalInSeconds)]);
+
+    // Assert
+    const latestBlock = await provider.getBlock("latest");
+    expect(latestBlock.number).to.equal(startingBlock.number + numberOfBlocks, "Block number mismatch");
+    expect(latestBlock.timestamp).to.equal(
+      startingTimestamp + (numberOfBlocks - 1) * intervalInSeconds * 1000 + 1,
+      "Timestamp mismatch"
+    );
+  });
+});
+
+describe("anvil_impersonateAccount & anvil_stopImpersonatingAccount", function () {
+  it("Should allow transfers of funds without knowing the Private Key", async function () {
+    // Arrange
+    const userWallet = Wallet.createRandom().connect(provider);
+    const beforeBalance = await provider.getBalance(RichAccounts[5].Account);
+
+    // Act
+    await provider.send("anvil_impersonateAccount", [RichAccounts[5].Account]);
+
+    const signer = await ethers.getSigner(RichAccounts[5].Account);
+    const tx = {
+      to: userWallet.address,
+      value: ethers.utils.parseEther("0.42"),
+    };
+
+    const recieptTx = await signer.sendTransaction(tx);
+    await recieptTx.wait();
+
+    await provider.send("anvil_stopImpersonatingAccount", [RichAccounts[5].Account]);
+
+    // Assert
+    expect((await userWallet.getBalance()).eq(ethers.utils.parseEther("0.42"))).to.true;
+    expect((await provider.getBalance(RichAccounts[5].Account)).eq(beforeBalance.sub(ethers.utils.parseEther("0.42"))))
+      .to.true;
+  });
+});
+
+describe("anvil_setCode", function () {
+  it("Should set code at an address", async function () {
+    // Arrange
+    const wallet = new Wallet(RichAccounts[0].PrivateKey);
+    const deployer = new Deployer(hre, wallet);
+
+    const address = "0x1000000000000000000000000000000000001111";
+    const artifact = await deployer.loadArtifact("Return5");
+    const contractCode = [...ethers.utils.arrayify(artifact.deployedBytecode)];
+
+    // Act
+    await provider.send("anvil_setCode", [address, contractCode]);
```

Comment:
afaiu, it's not a completely correct test. `hardhat_setCode`  does the same for this address. So it could be the case that this code will be already set to `0x1000000000000000000000000000000000001111 ` address

Reply:
Well, yes, random addresses should be more robust... Added one.